dd-java-agent/instrumentation/kotlin-coroutines/src/main/java/datadog/trace/instrumentation/kotlin/coroutines/ScopeStateCoroutineContext.java

```diff
@@ -1,27 +1,39 @@
 package datadog.trace.instrumentation.kotlin.coroutines;
 
+import static datadog.trace.instrumentation.kotlin.coroutines.CoroutineContextHelper.getJob;
+
 import datadog.trace.bootstrap.instrumentation.api.AgentScope;
 import datadog.trace.bootstrap.instrumentation.api.AgentTracer;
 import datadog.trace.bootstrap.instrumentation.api.ScopeState;
+import kotlin.Unit;
 import kotlin.coroutines.CoroutineContext;
+import kotlin.jvm.functions.Function1;
 import kotlin.jvm.functions.Function2;
 import kotlinx.coroutines.ThreadContextElement;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
-public class ScopeStateCoroutineContext implements ThreadContextElement<ScopeState> {
+public class ScopeStateCoroutineContext
+    implements ThreadContextElement<ScopeState>, Function1<Throwable, Unit> {
 
-  private static final Key<ScopeStateCoroutineContext> KEY = new ContextElementKey();
+  public static final Key<ScopeStateCoroutineContext> KEY = new ContextElementKey();
   private final ScopeState coroutineScopeState;
-  @Nullable private ContinuationHandler continuationHandler;
+  @Nullable private AgentScope.Continuation continuation;
+  @Nullable private AgentScope continuationScope;
+  private Boolean isInitialized = false;
 
   public ScopeStateCoroutineContext() {
     coroutineScopeState = AgentTracer.get().newScopeState();
-    final AgentScope activeScope = AgentTracer.get().activeScope();
-    if (activeScope != null) {
-      activeScope.setAsyncPropagation(true);
-      continuationHandler =
-          new ContinuationHandler(coroutineScopeState, activeScope.captureConcurrent());
+  }
+
+  public void maybeInitialize() {
+    if (!isInitialized) {
+      final AgentScope activeScope = AgentTracer.get().activeScope();
+      if (activeScope != null) {
+        activeScope.setAsyncPropagation(true);
```

Comment:
Also, we shouldn't force this on the active scope, but only capture it _iff_ it `isAsyncPropagating()`.

Reply:
Maybe we should make it an option? I.e. those who'd like to propagate scopes to coroutines by default will just enable this option, while others might have a more granular control over it?

Reply:
@monosoul Automatic async propagation is already the default. If the active scope has its async propagation set to `false` it is (hopefully) for a good reason. The other big switch will be the `kotlin_coroutine.experimental` name that enables/disables the integration completely, and it becomes available as `dd.trace.integration.kotlin_coroutine.experimental.enabled` and `DD_TRACE_INTEGRATION_KOTLIN_COROUTINE_EXPERIMENTAL_ENABLED`.

Reply:
Ahh, I see. Okay, cool, thanks for the explanation! :+1: